# **Instrucciones**
Elabore esta tarea en una notebook independiente, registrando sus resultados o las salidas de los programas. En todos los casos en que sea posible, resuelva los problemas mediante simulación y después verifique los resultados con las fórmulas correspondientes.


---
---


1. *En la biblioteca:*
Un encargado de la biblioteca es el único en el mostrador durante todo su turno de trabajo. Las llegadas al mostrador siguen una distribución de Poisson con una media de 8 por hora. Cada usuario de la biblioteca es atendido de uno en uno, y el tiempo de servicio sigue una distribución exponencial con una media de 5 minutos.

*   ¿Cuál es la probabilidad de que se forme cola?
*   ¿Cuál es la longitud media de la cola?
*   ¿Cuál es el tiempo medio que un cliente pasa en la biblioteca hasta que le han atendido?
*   ¿Cuál es el tiempo medio que un cliente pasa en la cola esperando a que le atiendan?
*   El encargado de la biblioteca pasa su tiempo en que no hay clientes clasificando artículos de revistas. Si puede clasificar 22 fichas por hora como media cuando trabaja continuamente, ¿cuántas fichas puede ordenar durante su trabajo?
---
---

In [6]:
from random import expovariate
import random
from statistics import mean
from math import inf as Infinity
from random import seed
from matplotlib.pyplot import *

def l_espera(lambd,my,a):

  Num_Pkts = 100000                   #Número de paquetes a simular
  count = 0                           #Contador de paquetes simulados
  t = 0
  N = 0                               #Variable de estado: número de paquetes en el sistema

  tll = expovariate(lambd)            #Tiempo de llegada
  ts = Infinity                       #Tiempo de salida
  t_evento_anterior = 0.0             #Tiempo del último evento

  tll_Data = []                       #Lista de tiempos de llegada
  ts_Data = []                        #Lista de tiempos de salida
  r_Data = []                         #Lista de retrasos individuales
  tf_Data=[]                          #Lista de tiempos en fila
  tc_Data =[]                         #Lista de tiempos en caja

  Data = {}

  while count < Num_Pkts:
      if tll < ts:                           #Evento de llegada
          t = tll
        
#Longitud entre intervalos (tiempo)
          delta = t - t_evento_anterior
          if N in Data: 
              Data[N] += delta
          else:
              Data[N] = delta
          t_evento_anterior = t
        
          tll_Data.append(t)
          N = N + 1.0
          tll = t + expovariate(lambd) 
          if N == 1:
              tc=expovariate(my)
              ts = t + tc
              tc_Data.append(tc)
      else:                                  #Evento de salida
          t = ts
        
#Longitud entre intervalos (tiempo)
          delta = t - t_evento_anterior
          if N in Data:
              Data[N] += delta
          else:
              Data[N] = delta
          t_evento_anterior = t
        
          ts_Data.append(t)
          N = N - 1.0
          count = count + 1                         #Paquetes simulados
          if N > 0:
              tc=expovariate(my)
              ts = t + tc
              tc_Data.append(tc)
          else:
              ts = Infinity

#DATOS DE SALIDA PROCESADOS
          
#Estimación de la probabilidad de **P(N=k)**
  for (key, value) in Data.items():
      Data[key] = value/t

  media = 0.0
  media2=0.0
  medial=[]
  for (key, value) in Data.items():

        media = media + key*value
        if (key-1)>=1:
          media2= media2 + (key-1.0)*value
        medial.append(value)
                                                                                
 #Opciones por realizar:
    #1) Retraso promedio 
    #2) Número medio de clientes en el sistema 
    #3) Probabilidad de que se forme en cola  

  if a==1:
#Estimación del retraso promedio:
    for i in range(Num_Pkts):
    #for i in range(T):
        d = ts_Data[i] - tll_Data[i]                #Retraso en sistema
        d2= ts_Data[i] - tll_Data[i]-tc_Data[i]     #Retraso en fila
        tf_Data.append(d2)
        r_Data.append(d)

    print( "Retraso promedio: W = ", round( mean(r_Data), 4) )
    print("Tiempo en fila: Wq =",mean(tf_Data))

  elif a==2:
    print('E[N]: L = ', media)
    print("L_q: ",media2)

  else:
    print('¿Cuál es la probabilidad de que se forme cola? p = ', sum(Data.values())-medial[0])
    print("¿Cuál es la probabilidad de que se encuentre vacio? q = ",medial[0])

---
Sabemos que nuestros parametros son: λ = 8, μ =12 .


A)¿Cuál es la probabilidad de que se forme cola?


$$ρ=\frac{λ}{μ}=\frac{8}{12}=\frac{2}{3}=0.666$$




In [8]:
print("Respuesta de la primer pregunta:\n ")
A=l_espera(8,12,3)

Respuesta de la primer pregunta:
 
¿Cuál es la probabilidad de que se forme cola? p =  0.6656595356674839
¿Cuál es la probabilidad de que se encuentre vacio? q =  0.33434046433251635


---
B) ¿Cuál es la longitud media de la cola?

$$L_q=λW_q=λ(W-\frac{1}{μ})=8(0.25-\frac{1}{12})=8(0.1698)=1.333$$


In [11]:
print("Respuesta de la segunda pregunta:\n ")
B_1=l_espera(8,12,1)
B_2=l_espera(8,12,2)

Respuesta de la segunda pregunta:
 
Retraso promedio: W =  0.2501
Tiempo en fila: Wq = 0.16663144115916625
E[N]: L =  2.021109552992929
L_q:  1.356702955968452


---
C) ¿Cuál es el tiempo medio que un cliente pasa en la biblioteca hasta que le han atendido?

Usando las salidas de programa, resolvemos manualmente:

$$W=W_q+\frac{1}{μ}=0.166+\frac{1}{12}=0.25$$


In [23]:
print("Respuesta de la tercera pregunta:\n ")
C=l_espera(8,12,1)

Respuesta de la tercera pregunta:
 
Retraso promedio: W =  0.2457
Tiempo en fila: Wq = 0.16273923416230598


---
D) ¿Cuál es el tiempo medio que un cliente pasa en la cola esperando a que le atiendan?

Usando las salidas de programa, resolvemos manualmente:

$$W_q=W-\frac{1}{μ}=0.25-\frac{1}{12}=0.1666$$


In [24]:
print("Respuesta de la cuarta pregunta:\n ")
D=l_espera(8,12,1)

Respuesta de la cuarta pregunta:
 
Retraso promedio: W =  0.2431
Tiempo en fila: Wq = 0.16033066569398075


---
E) El encargado de la biblioteca pasa su tiempo en que no hay clientes clasificando artículos de revistas. Si puede clasificar 22 fichas por hora como media cuando trabaja continuamente, ¿cuántas fichas puede ordenar durante su trabajo?

Usando las salidas de programa, resolvemos manualmente:

$$E(fichasporhora)=0.33(22)=7.26$$

In [25]:
print("Respuesta de la quinta pregunta:\n ")
D=l_espera(8,12,3)

Respuesta de la quinta pregunta:
 
¿Cuál es la probabilidad de que se forme cola? p =  0.6683781592366522
¿Cuál es la probabilidad de que se encuentre vacio? q =  0.3316218407633478


---
---
2. *Lavado de autos:*
Un pequeño autoservicio de lavado que solo tienen capacidad para atender de uno en uno a los autos que llegan, tiene un aparcamiento para 10 coches, incluyendo el que está siendo lavado. La empresa ha estimado que los coches llegan siguiendo una distribución de Poisson con una media de 20 coches/hora, el tiempo de servicio sigue una distribución exponencial
de 12 minutos. La empresa abre durante 10 horas al día.

*   ¿Cuál es la media de coches perdidos cada día debido a las limitaciones de espacio?
---
---
Sabemos que nuestros parametros son: λ = 20 coches/hora, μ =12 coches/hora y max = 10 horas al día.

In [29]:
def l_espera_2(lambd,my,max,total):

  T = total                           #Tiempo máximo 
  count = 0                           #Contador simulador
  t = 0
  N = 0                               #Variable de estado
  CNA=0                               #Clientes no atendidos

  tll =- ((1/lambd)*np.log(random.random()))       #Tiempo de llegada
  ts = Infinity                                    #Tiempo de salida
  t_evento_anterior = 0.0                          #Tiempo del último evento

  while t < T:
      if tll < ts:                     #Evento de llegada
          t = tll
        
#Longitud entre intervalos (tiempo)
          delta = t - t_evento_anterior
          t_evento_anterior = t
        
          if N>=max:                                    #Separa a los que no alcanzan a ser atendidos 
            CNA=CNA+1
 
          else:
            N = N + 1.0

          tll = t - ((1/lambd)*np.log(random.random())) 

          if N == 1:
              ts = t + expovariate(my)
      else:                                         #Evento de salida
          t = ts
        
#Longitud entre intervalos (tiempo)
          delta = t - t_evento_anterior
        
          N = N - 1.0
          count = count + 1                        #Paquetes simulados
          if N > 0:
              ts = t + expovariate(my)
          else:
              ts = Infinity
 
  return CNA

media = []

for i in range(10000):
  media.append(l_espera_2(20,5,10,10))

print("Respuesta del segundo problema:\n")
print("Promedio de autos sin atender= ",mean(media))

Respuesta del segundo problema:

Promedio de autos sin atender=  141.1636


---
---
3. *John Macko:*
John Macko es alumno en la U de Ozark. Hace trabajos extraños para aumentar sus ingresos. Las peticiones de trabajo llegan en promedio cada 5 días, pero el tiempo entre ellas es exponencial. El tiempo para terminar un trabajo también es exponencial, con una media de 4 días.
*   ¿Cuál es la probabilidad de que le falte trabajo a John?
*   Si John cobra unos 50 pesos por cada trabajo, ¿cuál es su ingreso mensual promedio?
*   Si al final del semestre John decide subcontratar los trabajos pendientes a 40 pesos cada uno, ¿cuánto debe esperar pagar en promedio?
---
---
Sabemos que nuestros parametros son: λ = 0.2 trabajos por día y μ =0.5 trabajos por día.

In [30]:
def l_espera_3(lambd,my,a,b,c,d):

  T = b
  count = 0                           #Contador de paquetes simulados
  t = 0
  N = 0                               #Variable de estado

  tll = expovariate(lambd)            #Tiempo de llegada
  ts = Infinity                       #Tiempo de salida
  t_evento_anterior = 0.0             #Tiempo del último evento

  tll_Data = []                       #Lista de tiempos de llegada
  ts_Data = []                        #Lista de tiempos de salida
  r_Data = []                         #Lista de retrasos individuales
  tc_Data =[]                         #Lista de tiempos en caja
  tf_Data=[]                          #Lista de tiempos en fila

  Data = {}

  while t < T:
      if tll < ts:                           # vento de llegada
          t = tll
        
#Longitud entre intervalos (tiempo)
          delta = t - t_evento_anterior
          if N in Data: 
              Data[N] += delta
          else:
              Data[N] = delta
          t_evento_anterior = t
        
          tll_Data.append(t)
          N = N + 1.0
          tll = t + expovariate(lambd) 
          if N == 1:
              tc=expovariate(my)
              ts = t + tc
              tc_Data.append(tc)
      else:                                         #Evento de salida
          t = ts
        
#Longitud entre intervalos (tiempo)
          delta = t - t_evento_anterior
          if N in Data:
              Data[N] += delta
          else:
              Data[N] = delta
          t_evento_anterior = t
        
          ts_Data.append(t)
          N = N - 1.0
          count = count + 1                         #Paquetes simulados
          if N > 0:
              tc=expovariate(my)
              ts = t + tc
              tc_Data.append(tc)
          else:
              ts = Infinity

#DATOS DE SALIDA PROCESADOS


#Estimación de la probabilidad de P(N=k):
  for (key, value) in Data.items():
      Data[key] = value/t

#Se verifica que la probabilidad total es 1.0

  media = 0.0
  media2=0.0
  medial=[]
  for (key, value) in Data.items():
                                                                                #Se verifican los valores de expectación 
        media = media + key*value
        if (key-1)>=1:                                                          #Se calcula lq
          media2= media2 + (key-1.0)*value
        medial.append(value)
                                                                                
#Opciones:
    #1) Retraso media 
    #2) Número medio de clientes en el sistema 
    #3) Probabilidad de que se forme cola  
    #4) Pago de los trabajo hechos 
    #5) Trabajadores para subcontratar   

  if a==1:
# Estimación del retraso promedio:
    for i in range(len(ts_Data)):
        d = ts_Data[i] - tll_Data[i]              #Retraso en sistema
        d2= ts_Data[i] - tll_Data[i]-tc_Data[i]   #Retraso en fila
        tf_Data.append(d2)
        r_Data.append(d)
    print( "Retraso promedio: W = ", round( mean(r_Data), 4) )
    print("Tiempo en fila:W_q=",mean(tf_Data))

  elif a==2:
    print('E[N]: L = ', media)
    print("L_q =",media2)

  elif a==3:
    print('¿Cuál es la probabilidad de que se forme cola? p = ', sum(Data.values())-medial[0])
    print("¿Cuál es la probabilidad de que se encuentre vacio? q = ",medial[0])

  elif a==4:
    return count*c

  else:
    return N*d

A) ¿Cuál es la probabilidad de que le falte trabajo a John?

$$p_0=1-p=1-\frac{λ}{μ}=1-\frac{0.2}{0.25}=0.2$$

In [32]:
A=l_espera_3(0.2,0.25,3,100,50,0)

¿Cuál es la probabilidad de que se forme cola? p =  0.9647601242230285
¿Cuál es la probabilidad de que se encuentre vacio? q =  0.03523987577697149


B) Si John cobra unos $50 por cada trabajo, ¿cuál es su ingreso mensual promedio?

In [35]:
media=[]
for i in range(10000):
  media.append(l_espera_3(0.2,0.25,4,30,50,0))
print("Su ingreso mensual promedio es: $",mean(media))

Su ingreso mensual promedio es: $ 229.04


C) Si al final del semestre John decide subcontratar los trabajos pendientes a $40 cada uno, ¿cuánto debe esperar pagar en promedio?

In [36]:
media=[]
for i in range(10000):
  media.append(l_espera_3(0.2,0.25,5,182.5,50,40))
print("El pago esperado promedio es: $",mean(media))

El pago esperado promedio es: $ 138.364
